### PR TITLE
[FEATURE][#178] Jinja templated with custom variables in Yaml file

### PR DESCRIPTION
**Pull request by elenamartina**
## Proposed changes
Change for Issue #178 

This allows to specify job-specific variables in a Yaml file for each deployment without having to export them as global environment variables (which, when doing multiple deployments at once, would be shared across different jobs and cause issues).

The deployment process has now a new optional field, `--template-variables-file`. When provided, an additional YAML file will be used to define custom variables to render Jinja2 deployment templates (default is just using environment variables).

## Types of changes

What types of changes does your code introduce to dbx?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Further comments
- Documentation needs to be updated

### CODE REVIEW

File: dbx/utils/common.py [hunk 1]
```diff
@@ -176,17 +176,32 @@ def get_all_environment_names(self) -> Any:
...
 
     def _render_jinja_template(self) -> str:
         path_list = self._path.split("/")
         file_name = path_list.pop()
         file_path = "/".join(path_list)
 
         j2_env = jinja2.Environment(loader=jinja2.FileSystemLoader(file_path))
-        return j2_env.get_template(file_name).render(os.environ)
+
+        if self._template_variables_file:
```
Comment (allebacco) [inline dbx/utils/common.py:191]: I think that the template rendering code could be moved outside the if-else, like:

```python
if self._template_variables_file:
    with open(self._template_variables_file, "r") as file:
        template_vars = ruamel.yaml.safe_load(file)
    template_vars = {**os.environ, **template_vars}
else:
    dbx_echo(
        "No custom template variables file provided."
        "Using only environment variables to render Jinja template."
    )
    template_vars = dict(os.environ)

rendered_template = j2_env.get_template(file_name).render(template_vars)
```

File: dbx/utils/common.py [hunk 2]
```diff
@@ -203,15 +218,18 @@ def get_all_environment_names(self) -> Any:
         return list(self._get_deployment_config().keys())
 
 
-def get_deployment_config(path: str) -> AbstractDeploymentConfig:
+def get_deployment_config(path: str, template_variables_file: str = None) -> AbstractDeploymentConfig:
```
Comment (allebacco) [inline dbx/utils/common.py:221]: `template_variables_file` type should be `typing.Optional[str]`

File: dbx/utils/common.py [hunk 3]
```diff
@@ -203,15 +217,18 @@ def get_all_environment_names(self) -> Any:
...
         return JsonDeploymentConfig(path)
     elif ext in ["yml", "yaml"]:
         return YamlDeploymentConfig(path)
     elif ext == "j2":
         second_ext = path.split(".")[-2]
-        return Jinja2DeploymentConfig(path, second_ext)
+        if template_variables_file:
+            return Jinja2DeploymentConfig(path, second_ext, template_variables_file)
+        else:
+            return Jinja2DeploymentConfig(path, second_ext)
```
Comment (mitchstockdale) [inline dbx/utils/common.py:231]: It looks like line 228-231 could be condensed into`return Jinja2DeploymentConfig(path, second_ext, template_variables_file)` given `None` default values in `Jinja2DeploymentConfig` and `get_deployment_config`

**Comment by renardeinside** [comment]
hi @elenamartina , 
LGTM - is it ready to be merged? :)

**Comment by elenamartina** [comment]
> hi @elenamartina , LGTM - is it ready to be merged? :)

Hi @renardeinside, it's good to go for me if it is for you!
The code coverage doesn't look too good even after I've added a test in `test_deploy`. Is that alright?

Also, some documentation should be added on the use of this new parameter. I've been quite busy so it'd be really great if you could!

**Comment by renardeinside** [comment]
> The code coverage doesn't look too good even after I've added a test in test_deploy. Is that alright?

I'm going to refactor the testing part anyways, so I don't see this as a major blocker :) 
yes, I'll take a look into it later this week.

**Comment by renardeinside** [comment]
hi @elenamartina , 
we're currently under heavy load due to dbx being added to [official](https://docs.databricks.com/dev-tools/dbx.html#) documentation and evolving into the main product. We'll need some time first to plan further development steps, we see a need for heavy refactoring and it requires some planning. I'll get back to this PR as soon as possible, hopefully in the next 2 weeks.

**Pull request merged**